app/recordtransfer/jobs.py

```diff
@@ -61,18 +61,23 @@ def create_downloadable_bag(submission: Submission, user_triggered: User) -> Non
                 new_job.save()
                 return
 
-        zipf = None
+        temp_zip_path = None
         try:
-            LOGGER.info("Zipping directory to an in-memory file ...")
-            zipf = BytesIO()
-            zipped_bag = zipfile.ZipFile(zipf, "w", zipfile.ZIP_DEFLATED, False)
-            zip_directory(submission.location, zipped_bag)
-            zipped_bag.close()
+            os.makedirs(settings.TEMP_STORAGE_FOLDER, exist_ok=True)
+            temp_fd, temp_zip_path = tempfile.mkstemp(
+                suffix=".zip", dir=settings.TEMP_STORAGE_FOLDER
+            )
+            os.close(temp_fd)
+
+            LOGGER.info("Creating zip file on disk at %s ...", temp_zip_path)
+            with zipfile.ZipFile(temp_zip_path, "w", zipfile.ZIP_DEFLATED, False) as zipped_bag:
+                zip_directory(submission.location, zipped_bag)
             LOGGER.info("Zipped directory successfully")
 
             file_name = f"{user_triggered.username}-{submission.bag_name}.zip"
             LOGGER.info("Saving zip file as %s ...", file_name)
-            new_job.attached_file.save(file_name, ContentFile(zipf.getvalue()), save=True)
+            with open(temp_zip_path, "rb") as f:
+                new_job.attached_file.save(file_name, ContentFile(f.read()), save=True)
```

Comment:
We still have the same memory issue here. The entirety of the file is read into memory with `read()`. Another issue is that you're opening the file again, whereas this is not needed.

Two things:

- `File` is the proper class to use here ([Django docs reference](https://docs.djangoproject.com/en/4.2/ref/files/file/#the-file-class))
- You can reuse `temp_zip_fd`, you just need to `seek()` back to the start of the file

```suggestion
            temp_zip_fd.seek(0)
            new_job.attached_file.save(file_name, File(temp_zip_fd), save=True)
```

Reply:
> We still have the same memory issue here. The entirety of the file is read into memory with `read()`. Another issue is that you're opening the file again, whereas this is not needed.
> 
> Two things:
> 
> * `File` is the proper class to use here ([Django docs reference](https://docs.djangoproject.com/en/4.2/ref/files/file/#the-file-class))
> * You can reuse `temp_zip_fd`, you just need to `seek()` back to the start of the file

That makes sense. I wonder whether `File` streams the contents of the file rather than read the entire file like I did. It's not mentioned anywhere in the docs explicitly, but I assume it does the former since I see it has a `chunks()` method that reads the file in chunks.

Since I'm using `tempfile.TemporaryFile`, which is a file-like object, I think I don't need to seek back to the start of the file before passing it to `File()`.

Reply:
> > We still have the same memory issue here. The entirety of the file is read into memory with `read()`. Another issue is that you're opening the file again, whereas this is not needed.
> > Two things:
> > 
> > * `File` is the proper class to use here ([Django docs reference](https://docs.djangoproject.com/en/4.2/ref/files/file/#the-file-class))
> > * You can reuse `temp_zip_fd`, you just need to `seek()` back to the start of the file
> 
> That makes sense. I wonder whether `File` streams the contents of the file rather than read the entire file like I did. It's not mentioned anywhere in the docs explicitly, but I assume it does the former since I see it has a `chunks()` method that reads the file in chunks.
> 
> Since I'm using `tempfile.TemporaryFile`, which is a file-like object, I think I don't need to seek back to the start of the file before passing it to `File()`.
> 
> Addressed by: [c22094c](https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/commit/c22094c163fd4a821d29877ca087d0c0385bc4da)

I went down a bit of a rabbit hole, but you're absolutely right about both your last points.

- In [File](https://github.com/django/django/blob/8f942f1c1dbf4222c8ca48253f7959366ed1bb60/django/core/files/base.py#L75), you can see that it calls `chunks()` when the file is iterated over.

- In [FieldFile](https://github.com/django/django/blob/8f942f1c1dbf4222c8ca48253f7959366ed1bb60/django/db/models/fields/files.py#L96), the file is passed through to the storage in `save`.

- In [FileSystemStorage](https://github.com/django/django/blob/8f942f1c1dbf4222c8ca48253f7959366ed1bb60/django/core/files/storage/filesystem.py#L121), you can see it reads the file chunk-by-chunk in `_save`.

- And back in [File](https://github.com/django/django/blob/8f942f1c1dbf4222c8ca48253f7959366ed1bb60/django/core/files/base.py#L55), you can see when `chunks()` is called, a `seek(0)` is always called every time the file is read

Reply:
That's perfect, thanks for looking into this!